File: 0901-1000/0940-Distinct Subsequences II/0940-Distinct Subsequences II.py
import collections
class Solution:
    def distinctSubseqII(self, S: str) -> int:
        total, end = 0, collections.Counter()
        MOD = 10 ** 9 + 7
        for c in S:
            total, end[c] = (total * 2 + 1 - end[c]) % MOD, (total + 1) % MOD
        return total

File: 0901-1000/0940-Distinct Subsequences II/0940-Distinct Subsequences II.cpp
class Solution {
public:
    int distinctSubseqII(string S) {
        long long total = 0;
        int MOD = 1e9 + 7;
        vector<int> end(26);
        for (char c : S) {
            long long temp = total;
            total = (total * 2 + 1 - end[c - 'a'] + MOD) % MOD;
            end[c - 'a'] = (temp + 1) % MOD;
        }
        return total;
    }
};
